spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/r2dbc/ConnectionFactoryBuilder.java

```diff
@@ -222,6 +251,22 @@ private String determineEmbeddedUsername(R2dbcProperties properties) {
 			}
 		}
 
+		private void configureIfProxyListenersExist(Builder optionsBuilder, ConnectionFactoryOptions urlOptions,
+				ProxyListenersProvider proxyListenersProvider) {
+			List<ProxyExecutionListener> proxyListeners = proxyListenersProvider.listeners();
+			if (proxyListeners.isEmpty()) {
+				return;
+			}
+			String driver = urlOptions.getRequiredValue(ConnectionFactoryOptions.DRIVER);
+			if (!ProxyConnectionFactoryProvider.PROXY_DRIVER.equals(driver)) {
+				optionsBuilder.option(ConnectionFactoryOptions.DRIVER, ProxyConnectionFactoryProvider.PROXY_DRIVER);
+				String protocol = urlOptions.getValue(ConnectionFactoryOptions.PROTOCOL);
+				String newProtocol = StringUtils.isEmpty(protocol) ? driver : driver + ":" + protocol;
+				optionsBuilder.option(ConnectionFactoryOptions.PROTOCOL, newProtocol);
+			}
+			optionsBuilder.option(ProxyConnectionFactoryProvider.PROXY_LISTENERS, proxyListeners);
+		}
```

Comment:
I am a bit confused what's going on there but it looks like the proxy can be created without having to wrap the `ConnectionFactory` and return it. In that case, I think the hook-point should provide you a way to customize the builder the way you're doing here. The fact the customization is actually registering proxy-related options should be an implementation detail.

Reply:
What here is doing is reconstructing `DRIVER` and `PROTOCOL` options.

e.g: `r2dbc:h2:mem://...` to `r2dbc:proxy:h2:mem://...`

| - | DRIVER | PROTOCOL |
| --- | --- | --- |
| From | h2 | mem |
| To| proxy | h2:mem |

This way, the discovery mechanism will create `ConnectionFactory` for proxy and mysql.

Actually, I think same way may be used for `r2dbc-pool` via `PoolingConnectionFactoryProvider` as well.